modules/message-lane/src/lib.rs

```diff
@@ -60,41 +61,45 @@ pub trait Trait<I = DefaultInstance>: frame_system::Trait {
 
 	/// They overarching event type.
 	type Event: From<Event<Self, I>> + Into<<Self as frame_system::Trait>::Event>;
-	/// Message payload.
-	type Payload: Parameter;
 	/// Maximal number of messages that may be pruned during maintenance. Maintenance occurs
 	/// whenever outbound lane is updated - i.e. when new message is sent, or receival is
 	/// confirmed. The reason is that if you want to use lane, you should be ready to pay
 	/// for it.
 	type MaxMessagesToPruneAtOnce: Get<MessageNonce>;
 
+	/// Payload type of outbound messages. This payload is dispatched on the bridged chain.
+	type OutboundPayload: Parameter;
+	/// Message fee type of outbound messages. This fee is paid on this chain.
+	type OutboundMessageFee: Parameter;
+
+	/// Payload type of inbound messages. This payload is dispatched on this chain.
+	type InboundPayload: Decode;
+	/// Message fee type of inbound messages. This fee is paid on the bridged chain.
+	type InboundMessageFee: Decode;
```

Comment:
We only need this to be able to decode the messages, right? I.e. it's not really used on the target chain?

Reply:
It isn't used now. But I had an idea (see "Some ideas for future PRs" of #339) on how to use it to build something like exchange rate between bridged chains weights/fee tokens. I.e. on the target chain you know:
1) weight `W` of the message;
2) fee `TF = WeightToFee(W)` in `TargetFeeTokens` that needs to be paid for this message;
3) fee `SF` in `SourceFeeTokens` that has been paid for delivery-and-dispatch of this message.

So if you have enough samples, you could (roughly) estimate conversion rate between `SourceFeeTokens` and `TargetFeeTokens` like that `ConversionRate = SourceFeeTokens / TargetFeeTokens`. This information could be used later in `LaneMessageVerifier::verify_message` on **this** chain to check if fee paid by the submitter is enough. So we may start (i.e. configure verifier) with some fixed rate and then it'll start slowly changing towards actual market rate.

So yes - not used now, but maybe used later.